utxo/src/undo.rs

```diff
@@ -100,20 +110,47 @@ impl BlockUndo {
     pub fn insert_tx_undo(
         &mut self,
         tx_id: Id<Transaction>,
-        tx_undo: TxUndo,
+        tx_undo: TxUndoWithSources,
     ) -> Result<(), BlockUndoError> {
         match self.tx_undos.entry(tx_id) {
-            std::collections::btree_map::Entry::Vacant(e) => e.insert(tx_undo),
-            std::collections::btree_map::Entry::Occupied(_) => {
-                return Err(BlockUndoError::UndoAlreadyExists(tx_id))
-            }
+            Entry::Vacant(e) => e.insert(TxUndo::new(tx_undo.utxos)),
+            Entry::Occupied(_) => return Err(BlockUndoError::UndoAlreadyExists(tx_id)),
         };
+
+        tx_undo
+            .sources
+            .into_iter()
+            .filter_map(|source_id| match source_id {
+                OutPointSourceId::Transaction(id) => Some(id),
+                OutPointSourceId::BlockReward(_) => None,
+            })
+            .for_each(|source_tx_id| {
+                self.dependencies.insert((tx_id, source_tx_id));
+            });
+
         Ok(())
     }
 
     pub fn take_tx_undo(&mut self, tx_id: &Id<Transaction>) -> Option<TxUndo> {
-        // TODO(PR): this should analyze dependencies before taking an undo
-        self.tx_undos.remove(tx_id)
+        // Check if the tx is a dependency for other txs.
+        let dependencies_count =
+            self.dependencies.iter().filter(|(_child, parent)| parent == tx_id).count();
+        if dependencies_count == 0 {
+            // If not this tx can be taken and returned.
+            // But first, remove itself as a dependency of others.
+            self.dependencies
+                .iter()
```

Comment:
Unfortunately, this is a bad design. This being O(N) basically means that undoing blocks is potentially O(N^2). That's a recipe for disaster with DoS attacks. We can't accept that. 

This should be at most O(log(N)). Meaning: You should provide a mechanism to search for transactions that doesn't involve looping over all transactions, but use some kind of tree at worst (like BTreeSet/Map).